common/src/main/java/com/wynntils/models/token/TokenModel.java

```diff
@@ -0,0 +1,252 @@
+            baking.type = type;
+            checkAndPromoteBaking();
+            return;
+        }
+
+        Matcher tokensMatcher = TOKEN_PATTERN.matcher(name);
+        if (tokensMatcher.matches()) {
+            CappedValue tokens =
+                    new CappedValue(Integer.parseInt(tokensMatcher.group(1)), Integer.parseInt(tokensMatcher.group(2)));
+
+            int id = event.getEntity().getId();
+            TokenGatekeeper gatekeeper = activeGatekeepers.get(id);
+            if (gatekeeper != null) {
+                // Update active gatekeeper
+                gatekeeper.setDeposited(tokens);
+                WynntilsMod.postEvent(new TokenGatekeeperEvent.Deposited(gatekeeper));
+                return;
+            }
+
+            TokenGatekeeper invisibleGatekeeper = invisibleGatekeepers.get(id);
+            if (invisibleGatekeeper != null) {
+                // We need to keep invisible gatekeepers up to date if they suddenly become visible
+                invisibleGatekeeper.setDeposited(tokens);
+                return;
+            }
+
+            // Create new baking gatekeeper
+            BakingTokenGatekeeper baking = getBaking(event.getEntity().position());
+            baking.value = tokens;
+            baking.valueEntityId = event.getEntity().getId();
+            checkAndPromoteBaking();
+            return;
+        }
+
+        // This does not provide any new information for the TokenGatekeeper, but it is a good
+        // extra check that we really match only on real gatekeepers
+        if (name.equals(VERIFICATION_STRING)) {
+            BakingTokenGatekeeper baking = getBaking(event.getEntity().position());
+            baking.confirmed = true;
+            checkAndPromoteBaking();
+        }
+    }
+
+    @SubscribeEvent
+    public void onLabelVisibility(EntityLabelVisibilityEvent event) {
+        if (!event.getVisibility()) {
+            // This is the normal way in which gatekeepers are "removed" when done
+            int id = event.getEntity().getId();
+            TokenGatekeeper gatekeeper = activeGatekeepers.get(id);
+            if (gatekeeper == null) return;
+
+            removeGatekeeper(id, gatekeeper);
+            invisibleGatekeepers.put(id, gatekeeper);
+        } else {
+            // It became visible again; restore it
+            int id = event.getEntity().getId();
+            TokenGatekeeper gatekeeper = invisibleGatekeepers.get(id);
+            if (gatekeeper == null) return;
+
+            invisibleGatekeepers.remove(id);
+            addGatekeeper(id, gatekeeper);
+        }
+    }
+
+    // This happens if we log out, switch class, etc
+    @SubscribeEvent
+    public void onEntitiesRemoved(RemoveEntitiesEvent event) {
+        for (int id : event.getEntityIds()) {
+            if (activeGatekeepers.containsKey(id)) {
+                removeGatekeeper(id, activeGatekeepers.get(id));
+            }
+            if (invisibleGatekeepers.containsKey(id)) {
```

Comment:
Qodana is right, no need for the contains

Reply:
I guess it is just me that think strange to rely on remove to work properly even for non-existing items. Oh well, I'll fix it if you want.